packages/material-ui-styles/src/withStyles/withStyles.test.js

```diff
@@ -235,6 +235,47 @@ describe('withStyles', () => {
       expect(sheetsRegistry.registry[0].rules.raw).to.deep.equal({ root: { padding: 9 } });
     });
 
+    it('should support the variants key', () => {
```

Comment:
This is a bit thin for testing. The current tests aren't sufficient in my opinion. Following that practice doesn't actually test anything but just covers lines.

I'd recommend not relying on any withStyles (or jss internals) and just writing tests that assert on the styles (skip in JSDOM).

That way we can actually test the matcher logic. Right now it's not apparent if the props in the matchers are OR or AND connected.

Reply:
I would also add a test for the new warning, as it's currently tight to the withStyle method that we might drop.

Reply:
> This is a bit thin for testing. The current tests aren't sufficient in my opinion. Following that practice doesn't actually test anything but just covers lines.
> 
> I'd recommend not relying on any withStyles (or jss internals) and just writing tests that assert on the styles (skip in JSDOM).
> 
> That way we can actually test the matcher logic. Right now it's not apparent if the props in the matchers are OR or AND connected.

@eps1lon should we then add test for the `getStylesCreator`? I saw that most of the test are done this way, that's why I added it here, but it makes sense to be extracted over there. I would avoid testing the logic for the `matcher` method as generally I would not like to test internals, so this seemed like a good enough approach (we have tests for testing the class keys generated for one and more keys, so that is covered already in the tests). In addition we have the visual test too. 

So do you think adding tests for the logic inside `getStylesCreator` is the correct approach?

> I would also add a test for the new warning, as it's currently tight to the withStyle method that we could? drop.

Will add one, let me just figure out the best place for testing it.

Reply:
> In addition we have the visual test too.

Right, forgot about argos: https://www.argos-ci.com/mui-org/material-ui/builds/2584.

I'd just render a component with a custom variant and the assert on the expected style e.g.

```js
render(<Button data-testid="button" variant="custom-variant" />);

const style = window.getComputedStyle(screen.getByTestId('button'));
expect(style).to.have.property('color', 'blue'); // or whatever custom styles you have
```

It's not necessarily about the happy path but about how different combinations interact e.g. a matcher only partially matches props.

Reply:
Makes sense, will add it

Reply:
Specificity might also be interesting i.e. does a matcher for `variant=outlined` and `size=small` override the styles for `variant=outlined`? How does this work with regards to specificity:
Do I have to rely on order of declaration or does each prop matched increase specificity?

Reply:
> Do I have to rely on order of declaration or does each prop matched increase specificity?

Currently it depends on order of definition, this probably is a good point for documentation (as well as the fact that the props matcher works with AND, not OR 👍

Reply:
I've added the tests in the `Button.test.js` file + added tests for warnings as well. I tried at first to make the test generic, but it depends on the other props in the component, so I decided to add it here..

Reply:
I don't even think that we should test this per component. Maybe we can just test this for one component under the `createMuiTheme.test.js` or some `variants.tests.js`. None of the logic is in the component's code, except relaxing the typings